Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs

```diff
@@ -1,20 +1,239 @@
+            }
+
+            ClickSound(chemMaster);
+        }
+
+        private void TransferReagents(ChemMasterComponent chemMaster, string reagentId, FixedPoint2 amount, bool fromBuffer)
+        {
+            var containerSlot = CompOrNull<ItemSlotsComponent>(chemMaster.Owner)?.Slots.GetValueOrDefault("container_slot");
+            if (containerSlot is null)
+                return;
+
+            if (!containerSlot.HasItem ||
+                containerSlot.Item is not {Valid: true} container ||
+                !TryComp(container, out FitsInDispenserComponent? fits) ||
+                !_solutionContainerSystem.TryGetSolution(container, fits.Solution, out var containerSolution)
+            )
+                return;
+
+            var bufferSolution = _solutionContainerSystem.EnsureSolution(chemMaster.Owner, ChemMasterComponent.SolutionName);
+
+            if (fromBuffer) // Buffer to container
+            {
+                amount = FixedPoint2.Min(amount, containerSolution.AvailableVolume);
+                amount = bufferSolution.RemoveReagent(reagentId, amount);
+                _solutionContainerSystem.TryAddReagent(container, containerSolution, reagentId, amount, out var _);
+            }
+            else // Container to buffer
+            {
+                amount = FixedPoint2.Min(amount, containerSolution.GetReagentQuantity(reagentId));
+                _solutionContainerSystem.TryRemoveReagent(container, containerSolution, reagentId, amount);
+                bufferSolution.AddReagent(reagentId, amount);
+            }
+
+            UpdateUiState(chemMaster, updateLabel: true);
+        }
+
+        private void DiscardReagents(ChemMasterComponent chemMaster, string reagentId, FixedPoint2 amount, bool fromBuffer)
+        {
+            var bufferSolution = _solutionContainerSystem.EnsureSolution(chemMaster.Owner, ChemMasterComponent.SolutionName);
+            var containerSlot = CompOrNull<ItemSlotsComponent>(chemMaster.Owner)?.Slots.GetValueOrDefault("container_slot");
+
+            if (fromBuffer)
+                bufferSolution.RemoveReagent(reagentId, amount);
+            else if (containerSlot is not null &&
+                containerSlot.HasItem &&
+                containerSlot.Item is {Valid: true} container &&
+                TryComp(container, out FitsInDispenserComponent? fits) &&
+                _solutionContainerSystem.TryGetSolution(container, fits.Solution, out var containerSolution))
+            {
+                _solutionContainerSystem.TryRemoveReagent(container, containerSolution, reagentId, amount);
+            }
+
+            UpdateUiState(chemMaster, updateLabel: fromBuffer);
+        }
+
+        private void OnCreatePillsMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterCreatePillsMessage message)
+        {
+            // Ensure the amount is valid.
+            if (message.Amount == 0 || message.Amount > chemMaster.PillProductionLimit)
+                return;
```

Comment:
like above this needs to check if its <0 as well because that would cause bugs

Reply:
Also an `uint`, can't be <0.

Reply:
true